File: matlab_files/python_substituted/stress_coupling.m
function [large_new_matrix] = stress_coupling(stress_matrices_paths, ...
    coupled_stress_folder, headers_on, new_stress_test, stress_headers)

dir_csv_stress_files = dir([stress_matrices_paths '\*.csv']);
csv_stress_files = {dir_csv_stress_files.name};
large_new_matrix = [];
large_new_matrix_path = [stress_matrices_path, '\Large_Stress_Matrix.csv'];

if exist(coupled_stress_folder)
    
else
    selected_component = input(['Enter the desired stress component(s) to modify,' ...
        ' possible values are Mises, S11, S22, S33, S12, S13, S23:\n']);
    selected_components = split(selected_component);
    selected_columns = zeros(length(selected_components),1);
    for k = 1:length(selected_components)
        if strcmp(selected_components{k}, 'Mises') == 1
            selected_columns(k) = 2;
        elseif strcmp(selected_components{k}, 'S11') == 1
            selected_columns(k) = 3;
        elseif strcmp(selected_components{k}, 'S22') == 1
            selected_columns(k) = 4;
        elseif strcmp(selected_components{k}, 'S33') == 1
            selected_columns(k) = 5;
        elseif strcmp(selected_components{k}, 'S12') == 1
            selected_columns(k) = 6;
        elseif strcmp(selected_components{k}, 'S13') == 1
            selected_columns(k) = 7;
        else
            selected_columns(k) = 8;
        end
    end
    
    for i = 1:length(csv_stress_files)
        size_to_evaluate = dir([dir_csv_stress_files(i).folder '\' csv_stress_files{i}]);
        if size_to_evaluate.bytes < 100
            csv_stress_files{i} = {};
        end
    end
    
    csv_stress_files = csv_stress_files(~cellfun(@isempty,csv_stress_files));
    
    for i = 1:length(csv_stress_files)
        
        %     equation_test = readmatrix([stress_matrices_paths '\' csv_stress_files{i}]);
        %     mises_stress = (1/sqrt(2))*sqrt((equation_test(:,3)-equation_test(:,4)).^2+...
        %         (equation_test(:,4)-equation_test(:,5)).^2 + (equation_test(:,5)-equation_test(:,3)).^2+...
        %         6.*(equation_test(:,6).^2+equation_test(:,8).^2+equation_test(:,7).^2))
        
        stresses_to_modify = readmatrix([stress_matrices_paths '\' csv_stress_files{i}]);
        new_stress_length = ones(length(stresses_to_modify),1);
        for j = 1:length(selected_columns)
            stresses_to_modify(:,selected_columns(j)) = stresses_to_modify(:,selected_columns(j))...
                + new_stress_length*new_stress_test;
        end
        new_matrix = stresses_to_modify;
        
        % Recalculate the Mises stress
        mises_stress = (1/sqrt(2))*sqrt((new_matrix(:,3)-new_matrix(:,4)).^2+...
            (new_matrix(:,4)-new_matrix(:,5)).^2 + (new_matrix(:,5)-new_matrix(:,3)).^2+...
            6.*(new_matrix(:,6).^2+new_matrix(:,8)+new_matrix(:,7).^2));
        new_matrix(:,2) = mises_stress;
        large_new_matrix = [large_new_matrix; new_matrix];
        if headers_on == 1
            new_matrix = [stress_headers; num2cell(new_matrix)];
            writecell(new_matrix,[coupled_stress_folder '\' csv_stress_files{i}]);
        else
            writematrix(new_matrix,[coupled_stress_folder '\' csv_stress_files{i}]);
        end
    end
    large_new_matrix = sortrows(large_new_matrix);
    if headers_on == 1
        large_new_matrix = [stress_headers; num2cell(large_new_matrix)];
        writecell(large_new_matrix, large_new_matrix_path);
    else
        writematrix(large_new_matrix, large_new_matrix_path);
    end
    new_e_column = large_new_matrix(1:end, 1);
    old_e_file_path = [base_path '\Iteration_' num2str(iteration - 1) 'e.csv'];
    e_file = readmatrix(old_e_file_path);
    e_file(1:end, end) = new_e_column;
    new_e_file_path = [base_path '\Iteration_' num2str(iteration) 'e.csv'];
    writematrix(e_file, new_e_file_path);
end
end

File: python_files_python_3/old/stress_coupling.py
from first_iter_actions import *


def stress_coupling(stress_matrices_path, coupled_stress_folder, headers_on, new_stress_test, stress_headers,
                    iteration, base_path, base_e_path, old_e_iter_path, new_e_iter_path):
    import os
    import numpy as np
    import csv
    import pdb

    # Copy what happens in MATLAB, and add the possibility of modifying the csv or dat file with Bdiff and Bdisl.
    # if iteration == 1:
    #     no_iter_path = first_iter_actions(base_path, base_e_path, new_e_iter_path)
    no_iter_path = first_iter_actions(base_path, base_e_path, new_e_iter_path)
    file_extension = 'csv'
    dir_csv_stress_files = os.listdir(stress_matrices_path)
    csv_stress_files = [filename for filename in dir_csv_stress_files if filename.endswith(file_extension)]
    large_new_matrix_path = os.path.join(stress_matrices_path, 'Large_Stress_Matrix.csv')
    large_new_matrix = []
    if os.path.exists(new_e_iter_path):
        pass
    else:
        selected_component = input('Enter the desired stress component(s) to modify, possible values are Mises, '
                                   'S11, S22, S33, S12, S13, S23:\n')
        selected_components = selected_component.split()
        selected_columns = np.zeros((len(selected_components), 1))
        for k in range(len(selected_components)):
            if selected_components[k] == 'Mises':
                selected_columns[k] = 1
            elif selected_components[k] == 'S11':
                selected_columns[k] = 2
            elif selected_components[k] == 'S22':
                selected_columns[k] = 3
            elif selected_components[k] == 'S33':
                selected_columns[k] = 4
            elif selected_components[k] == 'S12':
                selected_columns[k] = 5
            elif selected_components[k] == 'S13':
                selected_columns[k] = 6
            else:
                selected_columns[k] = 7

        for file_to_evaluate in range(len(csv_stress_files)):
            with open(os.path.join(stress_matrices_path, csv_stress_files[file_to_evaluate])) as matrix:
                matrix_to_evaluate = matrix.readlines()
                matrix_to_evaluate = [line.strip() for line in matrix_to_evaluate[1:]]
                matrix_to_evaluate = [np.array([eval(i) for i in line.split(",")[:]]) for line in matrix_to_evaluate]
                matrix_to_evaluate = np.array(matrix_to_evaluate)
                if 'Region' in csv_stress_files[file_to_evaluate]:
                    csv_stress_files[file_to_evaluate] = 'to_delete'
                if not matrix_to_evaluate[1:, 1:].any():
                    csv_stress_files[file_to_evaluate] = 'to_delete'
        csv_stress_files = filter(lambda a: a != 'to_delete', csv_stress_files)

        for i in range(len(csv_stress_files)):
            with open(os.path.join(stress_matrices_path, csv_stress_files[i])) as matrix:
                stresses_to_modify = matrix.readlines()
                stresses_to_modify = [line.strip() for line in stresses_to_modify[1:]]
                stresses_to_modify = [np.array([eval(elem) for elem in line.split(",")[:]])
                                      for line in stresses_to_modify]
                stresses_to_modify = np.array(stresses_to_modify)
                new_stress_length = np.ones((len(stresses_to_modify), 1))
                new_column = np.zeros((len(stresses_to_modify), ))
                for j in range(len(selected_columns)):
                    column_to_add = new_stress_length * new_stress_test
                    for k in range(len(stresses_to_modify[:, int(selected_columns[j])])):
                        new_column[k] = stresses_to_modify[k, int(selected_columns[j])] + int(column_to_add[k])
                    stresses_to_modify[:, int(selected_columns[j])] = new_column
                print matrix
                new_matrix = stresses_to_modify
                mises_stress = (1 / np.sqrt(2)) * np.sqrt((new_matrix[:, 2] - new_matrix[:, 3]) ** 2 +
                                                          (new_matrix[:, 3] - new_matrix[:, 4]) ** 2 +
                                                          (new_matrix[:, 4] - new_matrix[:, 2])
                                                          ** 2 + 6 * (new_matrix[:, 5] ** 2 + new_matrix[:, 7] +
                                                                      new_matrix[:, 6] ** 2))
                new_matrix[:, 1] = mises_stress
                large_new_matrix.append(new_matrix)
                print i
                print csv_stress_files[i]
                with open(os.path.join(coupled_stress_folder, csv_stress_files[i]), 'wb') as f_write:
                    if headers_on == 1:
                        writer = csv.writer(f_write)
                        writer.writerow(stress_headers)
                        writer.writerows(new_matrix)
                    else:
                        writer = csv.writer(f_write)
                        writer.writerows(new_matrix)

        # new_dictionary = {}
        # for new_matrix in large_new_matrix:
        #     for j in range(len(new_matrix)):
        #         dictionary_matrix = new_matrix[j]
        #         new_dictionary[dictionary_matrix[0]] = dictionary_matrix[1:]
        # new_ids = new_dictionary.keys()
        # sorted_matrix = np.zeros((len(new_ids), 8))
        # sorted_new_ids = list(sorted(new_ids))
        # for new in range(len(sorted_new_ids)):
        #     sorted_matrix[new, 0] = sorted_new_ids[new]
        #     sorted_matrix[new, 1:] = new_dictionary[sorted_new_ids[new]]
        # if headers_on == 1:
        #     with open(os.path.join(coupled_stress_folder, 'Large_stress_matrix.csv'), 'wb') as f_write:
        #         writer = csv.writer(f_write)
        #         writer.writerow(stress_headers)
        #         writer.writerows(sorted_matrix)
        # else:
        #     with open(os.path.join(coupled_stress_folder, 'Large_stress_matrix.csv'), 'wb') as f_write:
        #         writer = csv.writer(f_write)
        #         writer.writerows(sorted_matrix)

        # new_e_column = sorted_matrix[:, 1]
        # if iteration == 1:
        #     old_e_iter_path = no_iter_path
        # with open(old_e_iter_path, "r") as f:
        #     data = f.readlines()
        #     data = [line.strip() for line in data]
        #     data = [line.strip().replace(" ", ",") for line in data]
        #     data = [np.array([eval(i) for i in line.split(",")[:]]) for line in data]
        #     data = np.array(data)
        # data[:, -1] = new_e_column
        # data = np.matrix(data)
        # with open(new_e_iter_path, 'wb') as f_save:
        #     np.savetxt(f_save, data, fmt="%s")
        # with open(base_e_path, 'wb') as f_save:
        #     np.savetxt(f_save, data, fmt="%s")

        with open(os.path.join(coupled_stress_folder, 'EARTH.csv')) as matrix:
            e_stresses = matrix.readlines()
            e_stresses = [line.strip() for line in e_stresses[1:]]
            e_stresses = [np.array([eval(elem) for elem in line.split(",")[:]])
                          for line in e_stresses]
            e_stresses = np.array(e_stresses)

        new_e_column = e_stresses[:, 1]
        if iteration == 1:
            old_e_iter_path = no_iter_path
        with open(old_e_iter_path, "r") as f:
            data = f.readlines()
            data = [line.strip() for line in data]
            data = [line.strip().replace(" ", ",") for line in data]
            data = [np.array([eval(i) for i in line.split(",")[:]]) for line in data]
            data = np.array(data)
        data[:, -1] = new_e_column
        data = np.matrix(data)
        with open(new_e_iter_path, 'wb') as f_save:
            np.savetxt(f_save, data, fmt="%s")
        with open(base_e_path, 'wb') as f_save:
            np.savetxt(f_save, data, fmt="%s")

    # with open(new_e_iter_path, 'wb') as f_write:
    #     writer = csv.writer(f_write)
    #     writer.writerows(e_file)



            

